### PR TITLE
vertex ai vector search import & export

### PR DESCRIPTION
**Pull request by tottenjordan**
> implementation of import & export classes for Vertex AI Vector Search
> two notebooks to illustrate each class
> also a quickstart notebook if user doesn't have an index created already

### CODE REVIEW

**Comment by dhruv-anand-aintech** [comment]
Hi @tottenjordan, Thanks a lot for the comprehensive quickstart notebook!

I tried using the script on the command line, but got no vectors exported. Do more command line arguments need to be specified?

```bash
$ src/export_vdf.py vertexai_vectorsearch -
p onboarding-host-1d21888935424f -i projects/898052556820/locations/us-central1/indexes/7351462286659158016
Enter path to service account credentials file (hit return to use application default credentials): 
Optional: max_vectors to export; can be larger than actual vector count
Fetching indexes: 0it [00:00, ?it/s]
{
    "version": "0.0.8",
    "file_structure": [
        "vdf_20240131_211017_c6a75/VDF_META.json"
    ],
    "author": "dhruvanand",
    "exported_from": "vertexai_vectorsearch",
    "indexes": {}
}
Export to disk completed. Exported to: vdf_20240131_211017_c6a75/
Time taken to export data:  00:03:49
```

**Comment by tottenjordan** [comment]
Hi @dhruv-anand-aintech - absolutely, thanks for starting this! 

for current code, you should be able to supply just the index display name (as opposed to the `full resource name`)

I can either change this to just be the `full resource name`, or handle both?

**Comment by dhruv-anand-aintech** [comment]
> for current code, you should be able to supply just the index display name (as opposed to the `full resource name`)

I tried with `src/export_vdf.py vertexai_vectorsearch -p onboarding-host-1d21888935424f -i soverflow_vvs_vectorio_pubv3_20240131173351`, but got the same empty result. Maybe if the index name provided doesn't match any in the account, you can print that message out along with the possible options

**Comment by tottenjordan** [comment]
`soverflow_vvs_vectorio_pubv3_20240131173351` looks like the deployed index ID (i can tell because we added timestamp to naming convention, as this must be unique within a single project)

This is making it clear we should include handling for both!

For clarity, heres how to think  of the names/ids of index resources

`index.display_name` 
- example: "soverflow_vvs_vectorio_pubv3" [src](https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform/matching_engine/matching_engine_index.py#L120)

`index.resource_name` 
- example: "projects/934903580331/locations/us-central1/indexes/1081325705452584960" [like this](https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform/matching_engine/matching_engine_index.py#L62)

`my_index_endpoint.deployed_indexes[0].id` 
- example: "soverflow_vvs_vectorio_pubv3_20240130131739" [src](https://github.com/googleapis/python-aiplatform/blob/main/google/cloud/aiplatform/matching_engine/matching_engine_index_endpoint.py#L844)

user sets the `display_name` and can optional set the `deployed_index_id` else one will be auto generated. Resource name is given by GCP

will work on a commit asap to handle these!

File: notebooks/vertex_quickstart_w_bq_datasets.ipynb
Comment (dhruv-anand-aintech) [inline notebooks/vertex_quickstart_w_bq_datasets.ipynb:1]: not able to comment on the line, but I think `tree_ah_index.delete()` needs to be replaced with `my_vs_index.delete()`

**Comment by dhruv-anand-aintech** [comment]
I tried this:
```bash
src/import_vdf.py vertexai_vectorsearch -p onboarding-host-1d21888935424f -i soverflow_vvs_vectorio_pubv3
Enter the directory of vector dataset to be imported: vdf_20240125_130746_ac5a6_medium_articles
Enter region to host your index us-central1
Enter the Google Cloud Project Number 898052556820
size of upsert batches100
Optional. List of dicts for each datapoint; used to perform restricted searches
Optional. List of dicts for each datapoint;
Optional. CrowdingTag of the datapoint
list_of_ns_restrict_entries : None
list_of_numeric_entries : None
Traceback (most recent call last):
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 270, in <module>
    main()
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 257, in main
    import_vertexai_vectorsearch(args)
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 135, in import_vertexai_vectorsearch
    vertexai_vectorsearch_import = ImportVertexAIVectorSearch(args)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf/vertexai_vector_search_import.py", line 114, in __init__
    self.target_vertexai_index = aip.MatchingEngineIndex(self.target_index_resource_name)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/cloud/aiplatform/matching_engine/matching_engine_index.py", line 91, in __init__
    self._gca_resource = self._get_gca_resource(resource_name=index_name)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/cloud/aiplatform/base.py", line 646, in _get_gca_resource
    return getattr(self.api_client, self._getter_method)(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/cloud/aiplatform_v1/services/index_service/client.py", line 704, in get_index
    response = rpc(
               ^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 293, in retry_wrapped_func
    return retry_target(
           ^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 153, in retry_target
    _retry_error_helper(
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/retry/retry_base.py", line 212, in _retry_error_helper
    raise final_exc from source_exc
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/retry/retry_unary.py", line 144, in retry_target
    result = target()
             ^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 81, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InvalidArgument: 400 Request contains an invalid argument.
```

also tried `-i soverflow_vvs_vectorio_pubv3_20240131173351`

**Comment by tottenjordan** [comment]
@dhruv-anand-aintech - thanks for catching these, can you please try again?

note: 
- removed the `project_num` arg requirement
-  can pass either `deployed_index_id` or `index display_name` to both classes, import and export
- - if passing `display_name` to export class and that index is not deployed, will raise exception error informing user that only deployed indexes can be exported 

let me know if you see anything else to address

**Comment by dhruv-anand-aintech** [comment]
It'll also be good to set some good defaults for batch size, max vectors etc. Users can adjust via CLI arguments if required.

**Comment by dhruv-anand-aintech** [comment]
Hi @tottenjordan, I made some syntax and format fixes.

The import doesn't work without providing an index name:
```bash
$ src/import_vdf.py --dir vdf_20240125_130746_ac5a6_medium_articles vertexai_vectorsearch -p onboarding-host-1d21888935424f --location us-central1 
Enter the name of the index to import to:
Optional. Enter list of dicts describing string filters for each data point: 
Optional. Enter list of dicts for each datapoint: 
Optional. CrowdingTag of the datapoint: 
list_restrict_entries : None
list_of_numeric_entries : None
Traceback (most recent call last):
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 384, in <module>
    main()
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 371, in main
    import_vertexai_vectorsearch(args)
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 231, in import_vertexai_vectorsearch
    vertexai_vectorsearch_import = ImportVertexAIVectorSearch(args)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf/vertexai_vector_search_import.py", line 209, in __init__
    self.target_index = self._get_index()
                        ^^^^^^^^^^^^^^^^^
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf/vertexai_vector_search_import.py", line 310, in _get_index
    raise ResourceNotExistException("index")
import_vdf.vertexai_vector_search_import.ResourceNotExistException: Resource Does Not Exist.
```

Can it use the same index name (as prefix) as is provided in the vdf_ folder?
Example vdf folder to work with: https://huggingface.co/datasets/aintech/vdf_20240125_130746_ac5a6_medium_articles

**Comment by dhruv-anand-aintech** [comment]
> for the input arg statements, do you have any advice on how to decide what goes under `main()` vs `import_vertex(args)` vs both? for example, in the pinecone args, i see "subset" and "serverless" are only defined in `main()`, but are referenced in their import function to determine which set of args are used?

Usually each of the parameters would be present both as cli arguments via argparse in main(), and in the vectorDB's own export_/import_ function to be accepted with a prompt.

For subset, it is a top level switch, which asks for further details only if it is enabled, so the code is in the vector db specific function.

Ideally, we would want all the parameters to have reasonable default values/behaviors (eg: index name in import should be set based on what name is present in the VDF_META.json file of the dataset being imported.

File: src/import_vdf.py [hunk 1]
```diff
@@ -104,6 +105,133 @@ def import_pinecone(args):
...
+            type_name=str,
+            # required=True
+        )
+        set_arg_from_input(
+            args,
+            "dimensions",
+            "Required. The number of dimensions of the input vectors: ",
+            type_name=int,
+            # required=True
+        )
```
Comment (dhruv-anand-aintech) [inline src/import_vdf.py:152]: to be read per index from the VDF_META.json file

File: src/import_vdf.py [hunk 2]
```diff
@@ -165,6 +293,53 @@ def main():
     parser_qdrant = subparsers.add_parser(DBNames.QDRANT, help="Import data to Qdrant")
     parser_qdrant.add_argument("-u", "--url", type=str, help="Qdrant url")
 
+    # Vertex AI VectorSearch
+    parser_vertexai_vectorsearch = subparsers.add_parser(
+        DBNames.VERTEXAI, help="Import data to Vertex AI Vector Search"
+    )
+    parser_vertexai_vectorsearch.add_argument(
+        "-p", "--project-id", type=str, help="Google Cloud Project ID"
```
Comment (dhruv-anand-aintech) [inline src/import_vdf.py:301]: if not specified, we should provide a list to choose from (based on the logged in user's projects)

File: src/import_vdf.py [hunk 3]
```diff
@@ -104,6 +105,133 @@ def import_pinecone(args):
...
+        args, "crowding_tag", "Optional. CrowdingTag of the datapoint: ", type_name=str
+    )
+    if args["create_new_index"] is True:
+        set_arg_from_input(
+            args,
+            "gcs_bucket",
+            "Required. Valid gcs bucket name: ",
+            type_name=str,
+            # required=True
+        )
```
Comment (dhruv-anand-aintech) [inline src/import_vdf.py:145]: same here. it would be good to show the options, or create a new one if user hits enter

**Comment by tottenjordan** [comment]
> The import doesn't work without providing an index name:

ok so, is this understanding correct:

1. remove providing index_name in input args

2. it should grab index_name in the `self.vdf_meta["indexes"].items()`
 
3. does this imply if multiple index names were provided, and they don't exist, we always want to recreate the same number of indexes in Vertex, and mirror the import into each said index name?

**Comment by dhruv-anand-aintech** [comment]
Yes, this would be the expected behavior.
We also have a namespace level of partitioning possible under indexes.
For VectorDBs that don't support namespacing, we handle import by creating separate indexes for each namespace (eg: https://github.com/AI-Northstar-Tech/vector-io/blob/0b70b16fac5ff4131e0ea7e2375b45b1eec6fb08/src/import_vdf/qdrant_import.py#L36-L39) by using the format `{index_name}_{namespace_name}`.

**Comment by tottenjordan** [comment]
> We also have a namespace level of partitioning possible under indexes.
For VectorDBs that don't support namespacing, we handle import by creating separate indexes for each namespace

ok - for v1 I can adapt to the format specified above.

If i understand `namespaces` correctly:
1. We support deploying different indexes into the same endpoint. I think it's functionally similar to "namespace"
2. Additionally, the filter_restricts enable a "namespace" within an index (e.g. if you only want to filter against a `namespace` )

We can consider (1) and (2) and discuss if any changes should be made

File: src/import_vdf.py (hunk 3 above)
Comment (tottenjordan) [reply src/import_vdf.py:145]: got it - if user leaves blank will create new bucket in import class

**Comment by dhruv-anand-aintech** [comment]
Thanks for making the suggested changes. I think it is almost ready.

I tried running the import script:

```bash
src/import_vdf.py --dir vdf_20240125_130746_ac5a6_medium_articles vertexai_vectorsearch -p onboarding-host-1d21888935424f --location us-central1
```

and got:
```bash
Optional. Enter list of dicts describing string filters for each data point: 
Optional. Enter list of dicts for each datapoint: 
Optional. CrowdingTag of the datapoint: 
list_restrict_entries : None
list_of_numeric_entries : None
Importing data for index: medium_articles
index_meta: [{'namespace': '', 'index_name': 'medium_articles', 'total_vector_count': 5979, 'exported_vector_count': 5979, 'dimensions': 1536, 'model_name': 'text-embedding-ada-002', 'vector_columns': ['vector_text-embedding-ada-002_None', 'vector_text-embedding-3-small_4', 'vector_text-embedding-3-small_5', 'vector_text-embedding-3-small', 'title_vector'], 'data_path': 'medium_articles', 'metric': 'Euclid', 'model_map': {'vector_text-embedding-3-small': {'model_name': 'text-embedding-3-small', 'text_column': 'title', 'dimensions': 1536}, 'title_vector': {'model_name': 'facebook/dpr-ctx_encoder-single-nq-base', 'text_column': 'title'}, 'vector_text-embedding-3-small_5': {'model_name': 'text-embedding-3-small', 'text_column': 'title', 'dimensions': 5, 'vector_column': 'vector_text-embedding-3-small_5'}, 'vector_text-embedding-3-small_4': {'model_name': 'text-embedding-3-small', 'text_column': 'title', 'dimensions': 4, 'vector_column': 'vector_text-embedding-3-small_4'}, 'vector_text-embedding-ada-002_None': {'model_name': 'text-embedding-ada-002', 'text_column': 'title', 'dimensions': 1536, 'vector_column': 'vector_text-embedding-ada-002_None'}}}]
final_data_pathImporting to index: medium_articles
Full resource name: projects/898052556820/locations/us-central1/indexes/6306908648085913600
Target index config:
{
    "dimensions": 1536.0,
    "approximateNeighborsCount": 150.0,
    "distanceMeasureType": "DOT_PRODUCT_DISTANCE",
    "algorithmConfig": {
        "treeAhConfig": {
            "leafNodeEmbeddingCount": "1000",
            "leafNodesToSearchPercent": 7.0
        }
    },
    "shardSize": "SHARD_SIZE_MEDIUM"
}
Importing data from medium_articles to medium_articles
index_meta: [{'namespace': '', 'index_name': 'medium_articles', 'total_vector_count': 5979, 'exported_vector_count': 5979, 'dimensions': 1536, 'model_name': 'text-embedding-ada-002', 'vector_columns': ['vector_text-embedding-ada-002_None', 'vector_text-embedding-3-small_4', 'vector_text-embedding-3-small_5', 'vector_text-embedding-3-small', 'title_vector'], 'data_path': 'medium_articles', 'metric': 'Euclid', 'model_map': {'vector_text-embedding-3-small': {'model_name': 'text-embedding-3-small', 'text_column': 'title', 'dimensions': 1536}, 'title_vector': {'model_name': 'facebook/dpr-ctx_encoder-single-nq-base', 'text_column': 'title'}, 'vector_text-embedding-3-small_5': {'model_name': 'text-embedding-3-small', 'text_column': 'title', 'dimensions': 5, 'vector_column': 'vector_text-embedding-3-small_5'}, 'vector_text-embedding-3-small_4': {'model_name': 'text-embedding-3-small', 'text_column': 'title', 'dimensions': 4, 'vector_column': 'vector_text-embedding-3-small_4'}, 'vector_text-embedding-ada-002_None': {'model_name': 'text-embedding-ada-002', 'text_column': 'title', 'dimensions': 1536, 'vector_column': 'vector_text-embedding-ada-002_None'}}}]
Warning: More than one vector column found for index ['vector_text-embedding-ada-002_None', 'vector_text-embedding-3-small_4', 'vector_text-embedding-3-small_5', 'vector_text-embedding-3-small', 'title_vector']. Only the first vector column vector_text-embedding-ada-002_None will be imported.
vector_column_name    : vector_text-embedding-ada-002_None
vector_metadata_names : ['vector_text-embedding-ada-002_None', 'vector_text-embedding-3-small_4', 'vector_text-embedding-3-small_5', 'vector_text-embedding-3-small', 'title_vector']
Inserting data:   0%|              | 0/1 [00:16<?, ?it/s]
Traceback (most recent call last):
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 79, in error_remapped_callable
    return callable_(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/grpc/_channel.py", line 1160, in __call__
    return _end_unary_response_blocking(state, call, False, None)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/grpc/_channel.py", line 1003, in _end_unary_response_blocking
    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.RESOURCE_EXHAUSTED
        details = "The following quota exhausted: MatchingEngineStreamUpdateThroughput."
        debug_error_string = "UNKNOWN:Error received from peer ipv6:%5B2404:6800:4009:832::200a%5D:443 {created_time:"2024-02-06T01:57:55.690285+05:30", grpc_status:8, grpc_message:"The following quota exhausted: MatchingEngineStreamUpdateThroughput."}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 360, in <module>
    main()
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 347, in main
    import_vertexai_vectorsearch(args)
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf.py", line 218, in import_vertexai_vectorsearch
    vertexai_vectorsearch_import.upsert_data()
  File "/Users/dhruvanand/Code/vector-io/src/import_vdf/vertexai_vector_search_import.py", line 819, in upsert_data
    self.index_client.upsert_datapoints(request=upsert_request)
  File "/opt/homebrew/lib/python3.11/site-packages/google/cloud/aiplatform_v1/services/index_service/client.py", line 1161, in upsert_datapoints
    response = rpc(
               ^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
    return wrapped_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 81, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.ResourceExhausted: 429 The following quota exhausted: MatchingEngineStreamUpdateThroughput.
```

Can this be caught and perhaps the upserts frequency can be reduced (exp. backoff)?
Also, is there a way for me to increase this quota in my account?

**Comment by tottenjordan** [comment]
>   File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py", line 131, in __call__
>     return wrapped_func(*args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/opt/homebrew/lib/python3.11/site-packages/google/api_core/grpc_helpers.py", line 81, in error_remapped_callable
>     raise exceptions.from_grpc_error(exc) from exc
> google.api_core.exceptions.ResourceExhausted: 429 The following quota exhausted: MatchingEngineStreamUpdateThroughput.
> ```
> 
> Can this be caught and perhaps the upserts frequency can be reduced (exp. backoff)? 
> Also, is there a way for me to increase this quota in my account?

added `requests_per_minute` w/ default 6000

File: src/import_vdf/vertexai_vector_search_import.py [hunk 4]
```diff
@@ -0,0 +1,861 @@
...
+                for file in tqdm(parquet_files, desc="Iterating over parquet files"):
+                    file_path = os.path.join(final_data_path, file)
+                    df = pd.read_parquet(file_path)
+                    df["id"] = df["id"].apply(lambda x: str(x))
+
+                    insert_datapoints_payload = []
+
+                    for idx, row in tqdm(
+                        df.iterrows(), desc="Iterating over rows", total=len(df)
+                    ):
```
Comment (dhruv-anand-aintech) [inline src/import_vdf/vertexai_vector_search_import.py:756]: this loop can be rewritten to be more concise and efficient. Will leave it for you to do in a subsequent PR.

**Review by dhruv-anand-aintech — APPROVED** [review]
I was able to import my example dataset of around 6k vectors. Took a long time due to the rate limiting (~7x the time it took for milvus).

While trying to export the same index again, I hit some more quota limits, even though it didn't seem like I was using them.
I'll assume it works as before.
Screenshot 2024-02-06 at 1 51 51 PM
Screenshot 2024-02-06 at 2 05 30 PM

File: src/export_vdf/vertexai_vector_search_export.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+
+        dim = int(index_meta.get("metadata", {}).get("config", {}).get("dimensions", 0))
+
+        # start exporting
+        pbar = tqdm(total=total, desc=f"Exporting {index.display_name}")
+        # find nearest neighbors as proxy to export all datapoint ids
+        neighbors = index_endpoint.find_neighbors(
+            deployed_index_id=deployed_index_id,
+            queries=[[0.0] * dim],
+            num_neighbors=total,
```
Comment (dhruv-anand-aintech) [inline src/export_vdf/vertexai_vector_search_export.py:184]: Is there a max num_neighbors that can be queried. If so, what happens if total is more than that.

Comment (tottenjordan) [reply src/export_vdf/vertexai_vector_search_export.py:184]: No hard limit on num_neighbors to query

**Comment by tottenjordan** [comment]
> I was able to import my example dataset of around 6k vectors. Took a long time due to the rate limiting (~7x the time it took for milvus).

the downside of rate limiting... Should the default number be increased? if so, what value do you want? 

[1] What was `requests_per_minute` set to?
[2] were you able to increase your throughput / upsert quota?

**Comment by tottenjordan** [comment]
> While trying to export the same index again, I hit some more quota limits, even though it didn't seem like I was using them.

Do you already have a deployed index in region `us-central1`? Can check this and undeploy the indexes in the console, or through these commands:

(1) `gcloud ai index-endpoints list --project=$PROJECT_ID --region=us-central1`  (docs)
(2) `gcloud ai index-endpoints undeploy-index 456 --project=$PROJECT_ID --region=us-central1 --deployed-index-id=$DEPLOYED_INDEX_ID` ([docs](https://cloud.google.com/sdk/gcloud/reference/ai/index-endpoints/undeploy-index))

**Pull request merged**